diff '--color=auto' -u --recursive a/seat.c b/seat.c
--- a/seat.c
+++ b/seat.c
@@ -291,17 +291,6 @@
 		return true;
 	}
 #endif
-	if (server->allow_vt_switch && sym >= XKB_KEY_XF86Switch_VT_1 && sym <= XKB_KEY_XF86Switch_VT_12) {
-		if (wlr_backend_is_multi(server->backend)) {
-			if (server->session) {
-				unsigned vt = sym - XKB_KEY_XF86Switch_VT_1 + 1;
-				wlr_session_change_vt(server->session, vt);
-			}
-		}
-	} else {
-		return false;
-	}
-	wlr_idle_notifier_v1_notify_activity(server->idle, server->seat->seat);
 	return true;
 }
 
